pkg/controllers/management/clusterprovisioner/helm_cleaner.go

```diff
@@ -0,0 +1,297 @@
+				mappingItem.NewAPI = fmt.Sprintf(format, api.NewAPIVersion, api.Kind)
+			}
+
+			mappings.Mappings = append(mappings.Mappings, mappingItem)
+		}
+	}
+
+	return &mappings
+}
+
+func newClientGetter(k8sClient kubernetes.Interface, restConfig rest.Config) *wrangler.SimpleRESTClientGetter {
+	return &wrangler.SimpleRESTClientGetter{
+		ClientConfig:    &clientcmd.DefaultClientConfig,
+		RESTConfig:      &restConfig,
+		CachedDiscovery: memory.NewMemCacheClient(k8sClient.Discovery()),
+		RESTMapper:      meta.NewDefaultRESTMapper(nil),
+	}
+}
+
+func (p *Provisioner) cleanupHelmReleases(cluster *v3.Cluster) error {
+	clusterManager := p.clusterManager
+	userContext, err := clusterManager.UserContextNoControllers(cluster.Name)
+	if err != nil {
+		return errors.Wrapf(err, "[cleanupHelmReleases] failed to obtain the Kubernetes client instance")
+	}
+
+	clientGetter := newClientGetter(userContext.K8sClient, userContext.RESTConfig)
+
+	for _, namespace := range FeatureAppNS {
+		actionConfig := &action.Configuration{}
+		if err = actionConfig.Init(clientGetter, namespace, EmptyHelmDriverName, logrus.Infof); err != nil {
+			return errors.Wrapf(err, "[cleanupHelmReleases] failed to create ActionConfiguration instance for Helm")
+		}
+
+		listAction := action.NewList(actionConfig)
+		releases, err := listAction.Run()
+		if err != nil {
+			return errors.Wrapf(err, "[cleanupHelmReleases] failed to list Helm releases for namespace %v", namespace)
+		}
+
+		for _, helmRelease := range releases {
+			lastRelease, err := actionConfig.Releases.Last(helmRelease.Name)
+			if err != nil {
+				logrus.Errorf("[cleanupHelmReleases] failed to find latest release version for release %v", helmRelease.Name)
+				// If this fails, something went wrong. Skip to the next release.
+				continue
+			}
+
+			// TODO consume the function from helm-mapkubeapis once that is merged in
+			replaced, modifiedManifest, err := ReplaceManifestData(apiMappings, lastRelease.Manifest, cluster.Status.Version.GitVersion)
+			if err != nil {
+				// If this fails, it probably means we don't have adequate write permissions
+				return errors.Wrapf(err, "[cleanupHelmReleases] failed to replace deprecated/removed APIs on cluster %v", cluster.Name)
+			}
+
+			if !replaced {
+				logrus.Infof("[cleanupHelmReleases] release %v in namespace %v has no deprecated or removed APIs", lastRelease.Name, namespace)
+				continue
+			}
+
+			if err := updateRelease(lastRelease, modifiedManifest, actionConfig); err != nil {
+				return errors.Wrapf(err, "[cleanupHelmReleases] failed to update release %v in namespace %v", lastRelease.Name, lastRelease.Namespace)
+			}
+		}
+	}
+
+	return nil
+}
+
+// ReplaceManifestData replaces the out-of-date APIs with their respective valid successors, or removes an API that
+// does not have a successor.
+// Logic extracted from https://github.com/stormqueen1990/helm-mapkubeapis/blob/0245b7a7837a36fd164d83e496c453811d62c083/pkg/common/common.go#L81-L142
```

Comment:
I think we should fork rancher/mapkubeapis and put the code there, then do a go replace directive in our go.mod to point to our version. That way we don't need to maintain the code in Rancher's codebase as well.

Once the code has been merged upstream, we can then point back to upstream and archive our fork.

Reply:
I'm open to other perspectives, but I think our practices of maintaining our own forks is generally a bad one. We generally don't end up switching to the upstream and then we end up with another project that we need to patch/maintain (see client-go and helm). 

Given that the PR for helm-mapkubeapis hasn't merged yet, I'm ok with including the logic here, since this is the only project that we can say for sure that this will be in long-term.